// mqtts/dist/transport/websocket.transport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebsocketTransport = void 0;
const transport_1 = require("./transport");
const WebSocket = require("ws");
const stream_1 = require("stream");
const duplexify = require("duplexify");
class WebsocketTransport extends transport_1.Transport {
    constructor(options) {
        super(options);
        this.readable = new stream_1.PassThrough();
        this.writable = new stream_1.PassThrough();
        this.reset();
    }
    reset() {
        this.duplex = duplexify(this.writable, this.readable, { objectMode: true });
    }
    connect() {
        this.socket = new WebSocket(this.options.url, this.options.additionalOptions);
        this.socketStream = WebSocket.createWebSocketStream(this.socket, { objectMode: true });
        this.socketStream.pipe(this.readable);
        this.writable.pipe(this.socketStream);
        return new Promise(resolve => { var _a; return (_a = this.socket) === null || _a === void 0 ? void 0 : _a.on('open', resolve); });
    }
}
exports.WebsocketTransport = WebsocketTransport;
